mmverify.py

```diff
@@ -1,369 +1,588 @@
+        return any(tok in fr.c for fr in self)
+
+    def lookup_v(self, tok: Var) -> bool:
+        """Return whether the given token is an active variable."""
+        return any(tok in fr.v for fr in self)
+
+    def lookup_d(self, x: Var, y: Var) -> bool:
+        """Return whether the given ordered pair of tokens belongs to an
+        active disjoint variable statement.
+        """
+        return any((min(x, y), max(x, y)) in fr.d for fr in self)
+
+    def lookup_f(self, var: Var) -> Label:
+        """Return the label of the active floating hypothesis which types the
+        given variable.
+        """
+        for frame in self:
+            try:
+                return frame.f_labels[var]
+            except KeyError:
+                pass
         raise MMKeyError(var)
 
-    def lookup_d(self, x, y):
-        return any(((min(x,y), max(x,y)) in fr.d for fr in reversed(self)))
-
-    def lookup_e(self, stmt):
+    def lookup_e(self, stmt: Stmt) -> Label:
+        """Return the label of the (earliest) active essential hypothesis with
+        the given statement.
+        """
         stmt_t = tuple(stmt)
-        for frame in reversed(self):
-            try: return frame.e_labels[stmt_t]
-            except KeyError: pass
+        for frame in self:
+            try:
+                return frame.e_labels[stmt_t]
+            except KeyError:
+                pass
         raise MMKeyError(stmt_t)
 
-    def make_assertion(self, stat):
-        frame = self[-1]
+    def find_vars(self, stmt: Stmt) -> list[Var]:
+        """Return the list of variables in the given statement."""
+        var_list = []
+        for x in stmt:
+            if x not in var_list and self.lookup_v(x):
+                var_list.append(x)
+        return var_list
+
+    def make_assertion(self, stmt: Stmt) -> Assertion:
+        """Return a quadruple (disjoint variable conditions, floating
+        hypotheses, essential hypotheses, conclusion) describing the given
+        assertion.
+        """
         e_hyps = [eh for fr in self for eh in fr.e]
-        mand_vars = {tok for hyp in itertools.chain(e_hyps, [stat])
-                         for tok in hyp if self.lookup_v(tok)}
-
-        dvs = {(x,y) for fr in self for (x,y) in
+        mand_vars = {tok for hyp in itertools.chain(e_hyps, [stmt])
+                     for tok in hyp if self.lookup_v(tok)}
+        dvs = {(x, y) for fr in self for (x, y) in
                fr.d.intersection(itertools.product(mand_vars, mand_vars))}
-
-        f_hyps = collections.deque()
-        for fr in reversed(self):
-            for v, k in reversed(fr.f):
+        f_hyps = []
+        # If one allows Metamath databases with multiple $f-statements for a
+        # given var, then one should use "reversed" in the next two lines and
+        # use 'appendleft' from 'collections.deque' to get the latest f_hyp
+        # corresponding to the given var.
```

Comment:
Aren't multiple $f statements for the same variable forbidden _in the same scope_?
In any case I think the standard, or readability, recommends that in case a floating variable is declared at several places, it shall always be with the same typecode.

Reply:
You're right. I was trying too hard to justify the `reversed`s in the previous versions of the code.

(As for your second sentence, I would disagree for general mm databases. In many textbooks, it happens that, say, x denotes an element of an arbitrary group in one chapter, and then a real number in another chapter.)

Reply:
Regardless of our opinions toward it, I'm pretty sure that the second sentence is part of the spec.